### PR TITLE
Add crypto module for 2.5.x

### PR DESCRIPTION
**Pull request by wsargent**
A more ambitious Crypto module.

Breaks out Crypto into a trait and adds Guice bindings.
Deprecates AES methods (see #4407)
Adds more tests of signing, generating random methods.
Does a random.nextBytes() with 55 bytes, because why not.

### CODE REVIEW

**Comment by benmccann** [comment]
Maybe we could also add a documentation page demonstrating a use case for Play's signing methods so that people understand when and how to use them?

**Comment by wsargent** [comment]
@benmccann Good question.  This is going to be part of the blog post / announcement, but roughly speaking: the Crypto signing methods are useful in a very specific context, and not useful at all outside of it.  They are useful, mostly in situations where you want a user to employ a CookieBaker, so they shouldn't be deprecated... but...

The reason for having those signing functions is to provide decent authentication for session cookies.  In order to provide decent authentication for session cookies, Play may change the signing function from HMAC-SHA1.  It may happen because someone finds a second preimage attack against SHA1, it may happen simply because Blake2 is faster.  But when that happens, everyone who relies Crypto's signing function will break, which will suck.  The fewer people using Crypto.sign outside of cookie signing, the better.

There's another, more subtle reason for making the Crypto API private -- the "key separation principle" says that you should only ever use one key for one purpose.  In this case, `play.api.crypto.secret` is used for signing cookies.  We should not allow `play.api.crypto.secret` to be used for any other purpose -- not encryption, not signing using another algorithms, etc.  Unfortunately, this is one of those cryptographer folklore wisdoms that is both well-known in the field and completely unknown outside of it.

Finally, HMAC-SHA1 is available directly through JCA and (once you take out the auto-included secret key) there's no value being added by calling it through Play's Crypto API.

**Comment by benmccann** [comment]
@wsargent if the intention is for the signing methods to be private then perhaps you could deprecate them as well? You could make a version that is private that is used internally and by the public version, but have the public methods deprecated with a note to use `CookieBaker` instead

As you noted, changing the crypto signing methods could be painful. My app currently uses them and I would guess others do as well. So if the intention is for them not to be public and stable then it'd be nice to get 2.5 out with them deprecated before any big changes are made to them

**Comment by wsargent** [comment]
@benmccann That's... not a bad idea.  I'd need to make the CookieBaker take a Crypto object though.

File: framework/src/play/src/main/scala/play/api/libs/Crypto.scala [hunk 1]
```diff
@@ -278,10 +442,27 @@ case class CryptoConfig(
  *
  * https://www.owasp.org/index.php/Cryptographic_Storage_Cheat_Sheet
  */
-@Singleton
-class Crypto @Inject() (config: CryptoConfig) {
-
+class DefaultCrypto @Inject() (config: CryptoConfig) extends Crypto {
+
+  val clock = java.time.Clock.systemUTC()
```
Comment (gmethvin) [inline framework/src/play/src/main/scala/play/api/libs/Crypto.scala]: Could we make this clock injectable and bind it to systemUTC by default? I'm sure having a fake clock for testing would be useful in other places too.

Comment (wsargent) [inline framework/src/play/src/main/scala/play/api/libs/Crypto.scala]: Sent from Outlook

    _____________________________

**Pull request merged**